File: server/dao/credential.py
# Faraday Penetration Test IDE
# Copyright (C) 2016  Infobyte LLC (http://www.infobytesec.com/)
# See the file 'doc/LICENSE' for the license information

from sqlalchemy.orm.query import Bundle
from sqlalchemy import not_

from server.dao.base import FaradayDAO
from server.models import Credential, EntityMetadata
from server.utils.database import apply_search_filter


class CredentialDAO(FaradayDAO):

    MAPPED_ENTITY = Credential

    COLUMNS_MAP = {
        'couchid': [EntityMetadata.couchdb_id],
        'username': [Credential.username],
        'password': [Credential.password],
        'service_id': [],
        'host_id': []}

    STRICT_FILTERING = ["couchid"]

    def list(self, search=None, cred_filter={}):
        results = self.__query_database(search, cred_filter)

        rows = [self.__get_cred_data(result.cred) for result in results]
        result = {
            'rows': rows
        }

        return result

    def __query_database(self, search=None, cred_filter={}):
        creds_bundle = Bundle(
                'cred', Credential.username, Credential.password,
                Credential.name, EntityMetadata.couchdb_id,
                Credential.description, Credential.owned,
                EntityMetadata.revision, EntityMetadata.update_time,
                EntityMetadata.update_user, EntityMetadata.create_time,
                EntityMetadata.update_action, EntityMetadata.creator,
                EntityMetadata.update_controller_action, EntityMetadata.owner,
                EntityMetadata.command_id)

        query = self._session.query(creds_bundle)\
                             .outerjoin(EntityMetadata, EntityMetadata.id == Credential.entity_metadata_id)

        query = query.filter(Credential.workspace == self.workspace)
        # Apply filtering options to the query
        query = apply_search_filter(query, self.COLUMNS_MAP, search, cred_filter, self.STRICT_FILTERING)

        # I apply a custom filter for search by hostId and serviceId.
        # 'LIKE' for search by serviceId.%, that return only credentials started with serviceId.
        if cred_filter.get('service_id') is not None:
            query = query.filter(EntityMetadata.couchdb_id.like(cred_filter.get('service_id') + ".%"))

        # 'LIKE' for search by hostId.%, with that LIKE we receive credentials of services also.
        # I need another like for filter credentials of services (%.%.%)
        if cred_filter.get('host_id') is not None:
            query = query.filter(
                EntityMetadata.couchdb_id.like(cred_filter.get('host_id') + ".%")).filter(
                    not_(EntityMetadata.couchdb_id.like("%.%.%")))

        results = query.all()
        return results

    def __get_cred_data(self, cred):
        return {
            'id': cred.couchdb_id,
            'key': cred.couchdb_id,
            'value': {
                '_id': cred.couchdb_id,
                'username': cred.username,
                'password': cred.password,
                'owner': cred.owner,
                'owned': cred.owned,
                'description': cred.description,
                'name': cred.name,
                'metadata': {
                    'update_time': cred.update_time,
                    'update_user': cred.update_user,
                    'update_action': cred.update_action,
                    'creator': cred.creator,
                    'create_time': cred.create_time,
                    'update_controller_action': cred.update_controller_action,
                    'owner': cred.owner,
                    'command_id': cred.command_id
                },
                'couchid': cred.couchdb_id}}

File: server/www/scripts/credentials/providers/credential.js
// Faraday Penetration Test IDE
// Copyright (C) 2017  Infobyte LLC (http://www.infobytesec.com/)
// See the file 'doc/LICENSE' for the license information

"use strict";

angular.module('faradayApp')
    .factory('credential', ['ServerAPI', '$q', 
    function(ServerAPI, $q) {

        // All credentials need this properties minimum for build object.
        var _credentialFields = {
            'name': 'string',
            'username': 'string',
            'password': 'string',
            'type': 'string',
        };

        // Only this properties will be saved to server.
        var _credentialFieldsSaveToServer = {
            'name': 'string',
            'username': 'string',
            'metadata': 'string',
            'password': 'string',
            'type': 'string',
            'parent_type': 'string',
            'parent': 'string'
        };

        var Credential;
        Credential = function(data, parent, parent_type){
            if(data) {
                this.set(data, parent, parent_type);
            }
        };

        Credential.prototype = {
            // Build object.
            set: function(data, parent, parent_type) {

                data.type = 'Cred';
                data.parent_type = parent_type;
                data.parent = parent;
                if(data.metadata === undefined)
                    data.metadata = '';

                _checkFieldsOk(data);
                angular.extend(this, data);
            },

            // Find object in server and build that.
            load: function(ws, id){
                
                var deferred = $q.defer();
                var self = this;
                
                ServerAPI.getObj(ws, id, 'credential').then(function(response){
                    angular.extend(self, response.data);
                    deferred.resolve();
                });
                
                return deferred.promise;
            },

            // Delete object object in server.
            delete: function(ws) {
                return ServerAPI.deleteCredential(ws, this._id, this._rev);
            },

            // Update object in server.
            update: function(ws) {
                var self = this;
                self.metadata = updateMetadata(self.metadata);
                
                return ServerAPI.updateCredential(ws, buildObjectServer(self))
                .then(function(credentialData) {
                    self._rev = credentialData.rev;
                });
            },

            // Create object in server.
            create: function(ws) {
                var self = this;
                self.metadata = generateCreateMetadata();
                
                return ServerAPI.createCredential(ws, buildObjectServer(self)).
                    then(function(credential_data) {
                        self._rev = credential_data.rev;
                    });
            },

            getParentName: function(ws){
                
                var deferred = $q.defer();
                
                var result = this.parent;
                var hostIdToSearch = undefined;
                var serviceIdToSearch = undefined;

                //Parent is Host
                if (result.parent_type === 'Host'){
                    hostIdToSearch = result[0];

                    ServerAPI.getObj(ws, hostIdToSearch).then(function(response){
                        deferred.resolve(response.data.name);
                    });
                }

                //Parent is Service
                else if (result.lparent_type == 'Service'){
                    hostIdToSearch = result[0];
                    serviceIdToSearch = result.parent;

                     ServerAPI.getObj(ws, hostIdToSearch).then(function(responseHost){
                         ServerAPI.getObj(ws, serviceIdToSearch).then(function(responseService){
                            deferred.resolve(responseHost.data.name + '/' + responseService.data.name);
                         });
                    });
                }

                return deferred.promise;
            }
        };

    // Check object to construct have all fields and also, type of they are OK.
    // All fields in _credentialFields should are in object.
    var _checkFieldsOk = function(credential){

        Object.keys(_credentialFields).forEach(function(key, index) {
            // Credential dont have property or type of property in credential dont same.
            if(!credential.hasOwnProperty(key) || typeof(credential[key]) !== _credentialFields[key])
                throw 'Credential-Invalid fields: Invalid fields in credential creation: ' + key;
        });
    };

    // Build a credential object with only properties specified in _credentialFieldsSaveToServer (properties to save in server).
    var buildObjectServer =  function(credential){
        
        var serverObject = {};
        Object.keys(_credentialFieldsSaveToServer).forEach(function(key, index) {
             if(credential.hasOwnProperty(key))
                serverObject[key] = credential[key];
        });
        return serverObject;
    };

    var generateCreateMetadata = function() {

        return {
                'update_time': new Date().getTime(),
                'update_user': '',
                'update_action': 0,
                'creator': 'UI Web',
                'create_time': new Date().getTime(),
                'update_controller_action': '',
                'owner': ''
        };
    };

    var updateMetadata = function(metadata) {
        metadata['update_time'] =  new Date().getTime();
        return metadata;
    };

    return Credential;
}]);
